tower/src/builder/mod.rs

```diff
@@ -197,11 +197,23 @@ impl<L> ServiceBuilder<L> {
         f: F,
     ) -> ServiceBuilder<Stack<crate::util::MapRequestLayer<F>, L>>
     where
-        F: Fn(R1) -> R2,
+        F: FnMut(R1) -> R2 + Clone,
     {
         self.layer(crate::util::MapRequestLayer::new(f))
     }
 
+    /// Fallibly one request type to another, or to an error.
+    #[cfg(feature = "util")]
+    pub fn try_map_request<F, R1, R2, E>(
```

Comment:
The naming here feels like the unfortunate conjunction of two worlds... Usually `map` is infallible, and `and_then` or `then` is fallible, but we don't want this to happen *after* the service (implied by "service then closure"). There isn't anything like `before`... Maybe something like `filter` or `filter_request`?

Reply:
Hmm, I don't love this naming either. However, I'm not sure about `filter`, since we already have a [`Filter` middleware](https://github.com/tower-rs/tower/blob/master/tower/src/filter/mod.rs) which takes an async predicate...

Reply:
Uh huh... welp... I dunno, has this method been requested (with an example of need)? If not, my not-very-strong-opinion would be to just leave it off (it seems people could use `tower::filter` if needed?).

Reply:
It was added in https://github.com/tower-rs/tower/pull/396 --- @hlb8122, are you actually using `try_with` presently, and is there a motivation for using it instead of the `tower::filter` module?

It would be nice if we could add a `tower::filter::Predicate` impl for synchronous functions as well, but that would conflict with the blanket impl for functions returning `Future`:
https://github.com/tower-rs/tower/blob/704bfa819e3301e7aa0a35b6c09042d0b05670e4/tower/src/filter/predicate.rs#L15-L25

This feels like a case where the `futures` 0.1 `IntoFuture` trait could come in handy, by making it possible to automatically convert a synchronous `Result` into a ready future, so users don't have to wrap it. Then, the blanket impl could take `Fn(Request) -> IntoFuture<...>`...

Reply:
FWIW, I'm fine with removing the `try_map_request`/`try_with` methods and middleware entirely, too. It was just here when I sat down to start cleaning this up, so I erred on the side of keeping it.

Reply:
I'm not using `try_with`. I remember reading that @hdevalence was using some of the service combinators, maybe he'd like to comment on the usefulness perhaps.

A weakish defense of fallible mapping of requests would be: There might be some  transforms that are easier written as one transform using `try_with`/`try_map_request` rather than `with`/`map_request` then `filter`?

Reply:
That's a good point — a `try_map_request`/`try_with` allows changing the request type _and_ rejecting requests in one place, which the `tower::filter` interface doesn't. 

I guess the concern, though, is just that it doesn't feel great to have a bunch of different ways to do similar things that don't *quite* overlap. I think it would be good to have both synchronous and asynchronous filtering/fallible request mapping, but they should ideally have consistent API surfaces. 

I'm not opposed to the `filter` naming scheme for fallible-request-mapping APIs — it does seem much less wordy than `try_map_request`. maybe it would be good to have both synchronous and asynchronous versions of `filter` and expose them as combinators as well as layers?

Reply:
I think we should punt on this in *this* PR and address `filter`/`try_with` in a follow-up (see #502).